polars/polars-io/src/ndjson_core/ndjson.rs

```diff
@@ -0,0 +1,289 @@
+        let mut json_reader = CoreJsonReader::new(
+            reader_bytes,
+            self.n_rows,
+            self.schema,
+            self.n_threads,
+            1024, // sample size
+            self.chunk_size,
+            self.low_memory,
+            self.infer_schema_len,
+        )?;
+
+        let mut df: DataFrame = json_reader.as_df()?;
+        if rechunk && df.n_chunks()? > 1 {
+            df.as_single_chunk_par();
+        }
+        Ok(df)
+    }
+}
+
+pub(crate) struct CoreJsonReader<'a> {
+    reader_bytes: Option<ReaderBytes<'a>>,
+    n_rows: Option<usize>,
+    /// Explicit schema for the CSV file
+    schema: Cow<'a, Schema>,
+    n_threads: Option<usize>,
+    sample_size: usize,
+    chunk_size: usize,
+    low_memory: bool,
+}
+impl<'a> CoreJsonReader<'a> {
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new(
+        reader_bytes: ReaderBytes<'a>,
+        n_rows: Option<usize>,
+        schema: Option<&'a Schema>,
+        n_threads: Option<usize>,
+        sample_size: usize,
+        chunk_size: usize,
+        low_memory: bool,
+        infer_schema_len: Option<usize>,
+    ) -> Result<CoreJsonReader<'a>> {
+        let reader_bytes = reader_bytes;
+
+        let schema = match schema {
+            Some(schema) => Cow::Borrowed(schema),
+            None => {
+                let bytes: &[u8] = &reader_bytes;
+                let mut cursor = Cursor::new(bytes);
+
+                let data_type = ndjson::read::infer(&mut cursor, infer_schema_len).unwrap();
+                let schema: polars_core::prelude::Schema =
+                    StructArray::get_fields(&data_type).into();
+
+                Cow::Owned(schema)
+            }
+        };
+        Ok(CoreJsonReader {
+            reader_bytes: Some(reader_bytes),
+            schema,
+            sample_size,
+            n_rows,
+            n_threads,
+            chunk_size,
+            low_memory,
+        })
+    }
+    fn parse_json(&mut self, mut n_threads: usize, bytes: &[u8]) -> Result<DataFrame> {
+        let mut total_rows = 128;
+
+        if let Some((mean, std)) = get_line_stats(bytes, self.sample_size) {
+            total_rows = (bytes.len() as f32 / (mean - 0.01 * std)) as usize;
+        }
+
+        if total_rows == 128 {
+            n_threads = 1;
+        }
+
+        // let n_chunks = total_rows / self.chunk_size;
+
+        let rows_per_thread = total_rows / n_threads;
+
+        let max_proxy = bytes.len() / n_threads / 2;
+        let capacity = if self.low_memory {
+            self.chunk_size
+        } else {
+            std::cmp::min(rows_per_thread, max_proxy)
+        };
+
+        let expected_fields = &self.schema.len();
+        let file_chunks =
+            get_file_chunks(bytes, n_threads, *expected_fields, SEP, Some(QUOTE_CHAR));
+        // If the number of threads given by the user is lower than our global thread pool we create
+        // new one.
+        let owned_pool;
```

Comment:
You can just create less `file_chunks` in this case. If you send 2 units of works in `par_iter` only 2 threads will execute it.

Reply:
maybe im misunderstanding, are you suggesting to just use the `POOL.install` method instead of creating the task specific pool?

Reply:
Yes!